### PR TITLE
Support to create graph dynamically

### PR DESCRIPTION
**Pull request by Linary**
Implement #165

Change-Id: I03d3e606d6292acec46a7725094c05fe84a6c9cd

### CODE REVIEW

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 1]
```diff
@@ -54,19 +55,26 @@
...
 
+    // TODO: any graceful way to record graphsDir?
+    private final String graphsDir;
     private final Map<String, Graph> graphs;
     private final HugeAuthenticator authenticator;
+    private final EventHub eventHub;
 
-    public GraphManager(HugeConfig conf) {
+    public GraphManager(HugeConfig conf, String graphsDir, EventHub hub) {
+        this.graphsDir = graphsDir;
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:74]: prefer to read from conf

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 2]
```diff
@@ -184,7 +203,6 @@ private void loadGraph(String name, String path) {
         final Graph graph = GraphFactory.open(path);
         this.graphs.put(name, graph);
         LOG.info("Graph '{}' was successfully configured via '{}'", name, path);
-
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:187]: keep the line

File: hugegraph-api/src/main/java/com/baidu/hugegraph/server/RestServer.java [hunk 3]
```diff
@@ -69,6 +78,17 @@ public void start() throws IOException {
         this.calcMaxWriteThreads();
     }
 
+    private void fillGraphsOption(String graphsDir) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/server/RestServer.java:81]: move to graph manager

File: hugegraph-api/src/main/java/com/baidu/hugegraph/server/RestServer.java [hunk 4]
```diff
@@ -69,6 +78,17 @@ public void start() throws IOException {
...
+        Map<String, String> graphConfs = ConfigUtil.scanGraphsDir(graphsDir);
+        List<String> optionValues = new ArrayList<>(graphConfs.size());
+        for (Map.Entry<String, String> entry : graphConfs.entrySet()) {
+            String optionValue = entry.getKey() + ":" + entry.getValue();
+            optionValues.add(optionValue);
+        }
+        this.conf.setProperty(ServerOptions.GRAPHS.name(),
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/server/RestServer.java:88]: append graphs and keep the origin graphs

File: hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java [hunk 5]
```diff
@@ -126,6 +126,7 @@ public static synchronized ServerOptions instance() {
                     2 * Runtime.getRuntime().availableProcessors()
             );
 
+    // TODO: Do not allow users to specify
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/config/ServerOptions.java:129]: seems can keep compatible

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/cmd/InitStore.java [hunk 6]
```diff
@@ -20,28 +20,26 @@
...
+import com.baidu.hugegraph.util.ConfigUtil;
 import com.baidu.hugegraph.util.E;
 import com.baidu.hugegraph.util.Log;
 
 public class InitStore {
 
     private static final Logger LOG = Log.logger(InitStore.class);
 
-    private static final String GRAPHS = ServerOptions.GRAPHS.name();
+    private static final String GRAPHS = "graphs";
```
Comment (javeme) [inline hugegraph-dist/src/main/java/com/baidu/hugegraph/cmd/InitStore.java:42]: still keep it

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 7]
```diff
@@ -88,6 +96,17 @@ public void loadGraphs(final Map<String, String> graphConfs) {
         }
     }
 
+    public HugeGraph createGraph(HugeConfig config) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:99]: ensure auth permission: write graph

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java [hunk 8]
```diff
@@ -33,17 +33,42 @@
...
+
+    private void listenChanges() {
+        this.hub.listen(Events.GRAPH_CREATE, event -> {
+            event.checkArgs(HugeGraph.class);
+            Object[] args = event.args();
+            HugeGraph graph = (HugeGraph) args[0];
+            this.injectGraph(graph);
+            return null;
+        });
+    }
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java:63]: unlistenChanges at close

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/cmd/InitStore.java (hunk 6 above)
Comment (Linary) [reply hugegraph-dist/src/main/java/com/baidu/hugegraph/cmd/InitStore.java:42]: move to ConfigUtil

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 9]
```diff
@@ -96,14 +98,46 @@ public void loadGraphs(final Map<String, String> graphConfs) {
         }
     }
 
-    public HugeGraph createGraph(HugeConfig config) {
+    public HugeGraph cloneGraph(String name, GraphsAPI.JsonGraphParams params) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:101]: pass the newName instead of JsonGraphParams

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 10]
```diff
@@ -96,14 +98,46 @@ public void loadGraphs(final Map<String, String> graphConfs) {
...
+
+        HugeGraph g = graph(name);
+        HugeConfig config = (HugeConfig) g.configuration();
+        HugeConfig cloneConfig = (HugeConfig) config.clone();
+        cloneConfig.setDelimiterParsingDisabled(true);
+        cloneConfig.setProperty(CoreOptions.STORE.name(), params.name());
+        return this.createGraph(cloneConfig);
+    }
+
+    public HugeGraph createGraph(GraphsAPI.JsonGraphParams params) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:120]: pass name and options parameters

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java [hunk 11]
```diff
@@ -82,7 +84,7 @@ public Object list(@Context GraphManager manager,
     @Timed
     @Path("{name}")
     @Produces(APPLICATION_JSON_WITH_CHARSET)
-    @RolesAllowed({"admin", "$owner=$name"})
+    @RolesAllowed({"admin"})
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java:87]: keep owner

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java [hunk 12]
```diff
@@ -91,6 +93,40 @@ public Object get(@Context GraphManager manager,
...
+            LOG.debug("Create graph {} with copied config from '{}'",
+                      params.name, name);
+            HugeGraph graph = manager.cloneGraph(name, params);
+            return ImmutableMap.of("name", graph.name(),
+                                   "backend", graph.backend());
+        } else if (action.equals("drop")) {
+            LOG.debug("Drop graph} by name '{}'", name);
+            throw new UnsupportedOperationException("");
+        }
+        return null;
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java:117]: E.checkArgument(false,...)

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java [hunk 13]
```diff
@@ -91,6 +93,40 @@ public Object get(@Context GraphManager manager,
...
+                        "The param action can't be null");
+        if (action.equals("clone")) {
+            LOG.debug("Create graph {} with copied config from '{}'",
+                      params.name, name);
+            HugeGraph graph = manager.cloneGraph(name, params);
+            return ImmutableMap.of("name", graph.name(),
+                                   "backend", graph.backend());
+        } else if (action.equals("drop")) {
+            LOG.debug("Drop graph} by name '{}'", name);
+            throw new UnsupportedOperationException("");
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java:115]: implement it

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java [hunk 14]
```diff
@@ -91,6 +93,40 @@ public Object get(@Context GraphManager manager,
         return ImmutableMap.of("name", g.name(), "backend", g.backend());
     }
 
+    @PUT
+    @Timed
+    @Path("{name}")
+    @Produces(APPLICATION_JSON_WITH_CHARSET)
+    @RolesAllowed({"admin"})
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java:100]: keep owner

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 15]
```diff
@@ -88,6 +98,49 @@ public void loadGraphs(final Map<String, String> graphConfs) {
         }
     }
 
+    public HugeGraph cloneGraph(String name, GraphsAPI.JsonGraphParams params) {
+        /*
+         * 0. check and modify params
+         * 1. init backend store
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:104]: add "create graph instance"

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 16]
```diff
@@ -88,6 +98,49 @@ public void loadGraphs(final Map<String, String> graphConfs) {
...
+        config.setProperty(CoreOptions.STORE.name(), params.name());
+        return this.createGraph(config);
+    }
+
+    private HugeGraph createGraph(HugeConfig config) {
+        HugeGraph graph = (HugeGraph) GraphFactory.open(config);
+        graph.initBackend();
+        // Inject into gremlin server context
+        this.eventHub.call(Events.GRAPH_CREATE, graph);
+        this.graphs.put(graph.name(), graph);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:138]: also add comment

File: hugegraph-dist/src/assembly/static/conf/gremlin-server.yaml [hunk 17]
```diff
@@ -6,9 +6,9 @@
 scriptEvaluationTimeout: 30000
 
 channelizer: org.apache.tinkerpop.gremlin.server.channel.WsAndHttpChannelizer
-graphs: {
-  hugegraph: conf/hugegraph.properties
-}
+#graphs: {
+#  hugegraph: conf/graphs/hugegraph.properties
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/conf/gremlin-server.yaml:10]: keep "graphs" key and empty value

File: hugegraph-dist/src/assembly/static/conf/rest-server.properties [hunk 18]
```diff
@@ -3,8 +3,7 @@ restserver.url=http://127.0.0.1:8080
 # gremlin server url, need to be consistent with host and port in gremlin-server.yaml
 #gremlinserver.url=http://127.0.0.1:8182
 
-# graphs list with pair NAME:CONF_PATH
-graphs=[hugegraph:conf/hugegraph.properties]
+graphs=conf/graphs
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/conf/rest-server.properties:6]: ./conf/graphs

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java [hunk 19]
```diff
@@ -91,6 +97,45 @@ public Object get(@Context GraphManager manager,
...
+        }
+    }
+
+    @POST
+    @Timed
+    @Produces(APPLICATION_JSON_WITH_CHARSET)
+    @RolesAllowed({"admin"})
+    public Object create(@Context GraphManager manager,
+                         JsonGraphParams params) {
+        HugeGraph graph = manager.createGraph(params.name, params.options);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java:135]: add LOG.debug

File: hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java [hunk 20]
```diff
@@ -33,17 +34,55 @@
...
         super(settings, newGremlinExecutorService(settings));
+        this.hub = hub;
+        this.listenChanges();
+    }
+
+    private void listenChanges() {
+        this.hub.listen(Events.GRAPH_CREATE, event -> {
+            event.checkArgs(HugeGraph.class);
+            Object[] args = event.args();
+            HugeGraph graph = (HugeGraph) args[0];
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/auth/ContextGremlinServer.java:62]: keep consistent with drop: event.args()

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java [hunk 21]
```diff
@@ -90,18 +106,20 @@ public void stop() {
...
         }
 
-        HugeRestServer.register();
+        HugeGraphServer.register();
 
         HugeGraphServer server = new HugeGraphServer(args[0], args[1]);
         Runtime.getRuntime().addShutdownHook(new Thread(() -> {
             LOG.info("HugeGraphServer stopping");
             server.stop();
-            LOG.info("HugeGraphServer stopped");
```
Comment (javeme) [inline hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java:104]: sure to remove it?

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGremlinServer.java [hunk 22]
```diff
@@ -25,60 +25,34 @@
...
-
         try {
             settings = Settings.read(conf);
         } catch (Exception e) {
             LOG.error("Can't found the configuration file at {} or " +
                       "being parsed properly. [{}]", conf, e.getMessage());
             throw e;
         }
+        // Scan graph confs and inject into gremlin server context
+        settings.graphs.putAll(ConfigUtil.scanGraphsDir(graphsDir));
```
Comment (javeme) [inline hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGremlinServer.java:49]: check settings.graphs is empty before put

File: hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java (hunk 21 above)
Comment (Linary) [reply hugegraph-dist/src/main/java/com/baidu/hugegraph/dist/HugeGraphServer.java:104]: not sure, just feel the logs are a bit redundant

**Comment by javeme** [comment]
api test error:
```
# Set backend and serializer
sed -i "s/backend=.*/backend=$BACKEND/" $CONF
sed: can't read hugegraph-0.11.1/conf/hugegraph.properties: No such file or directory
cat: hugegraph-0.11.1/logs/hugegraph-server.log: No such file or directory
The command "$TRAVIS_DIR/run-api-test.sh" exited with 1.
312.45s$ $TRAVIS_DIR/run-unit-test.sh
```

**Comment by javeme** [comment]
cassandra:
```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.20:test (core-test) on project hugegraph-test: There are test failures.
[ERROR] 
[ERROR] Please refer to /home/travis/build/hugegraph/hugegraph/hugegraph-test/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date]-jvmRun[N].dump, [date].dumpstream and [date]-jvmRun[N].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :hugegraph-test
The command "mvn test -P core-test,$BACKEND" exited with 1.
```

postgre:
```
[ERROR] Failed to execute goal org.jacoco:jacoco-maven-plugin:0.8.4:dump (pull-test-data) on project hugegraph-test: Unable to dump coverage data: Connection refused (Connection refused) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
The command "$TRAVIS_DIR/run-api-test.sh" exited with 1.
```

File: hugegraph-dist/src/assembly/static/conf/gremlin-server.yaml [hunk 23]
```diff
@@ -68,7 +68,7 @@ serializers:
   }
 metrics: {
   consoleReporter: {enabled: false, interval: 180000},
-  csvReporter: {enabled: true, interval: 180000, fileName: /tmp/gremlin-server-metrics.csv},
+  csvReporter: {enabled: false, interval: 180000, fileName: /tmp/gremlin-server-metrics.csv},
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/conf/gremlin-server.yaml:71]: submit an independent commit, and change `/tmp/` to `./metrics/`

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 24]
```diff
@@ -98,6 +111,114 @@ public void waitGraphsStarted() {
...
+        E.checkArgument(this.graphs.size() > 1,
+                        "The graph '%s' is the only one, not allowed to delete",
+                        name);
+        g.clearBackend();
+        try {
+            g.close();
+        } catch (Exception e) {
+            LOG.warn("Failed to close graph", e);
+        }
+        // Inject into gremlin server context
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:213]: update comment

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 25]
```diff
@@ -98,6 +111,114 @@ public void waitGraphsStarted() {
...
+        this.eventHub.call(Events.GRAPH_DROP, name);
+        // Inject into rest server context
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:215]: ditto

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 26]
```diff
@@ -98,6 +111,114 @@ public void waitGraphsStarted() {
...
+        E.checkArgument(!this.graphs().contains(name),
+                        "The graph name '%s' has existed", name);
+
+        PropertiesConfiguration propConfig = this.buildConfig(configText);
+        HugeConfig config = new HugeConfig(propConfig, false);
+        this.checkOptions(config);
+        return this.createGraph(config);
+    }
+
+    private PropertiesConfiguration buildConfig(String configText) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:152]: move to line 189

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 27]
```diff
@@ -98,6 +111,114 @@ public void waitGraphsStarted() {
...
+        try {
+            graph.initBackend();
+        } catch (BackendException e) {
+            HugeFactory.remove(graph);
+            throw e;
+        }
+        // Inject into gremlin server context
+        this.eventHub.call(Events.GRAPH_CREATE, graph);
+        // Inject into rest server context
+        this.graphs.put(graph.name(), graph);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:184]: prefer to listen event and don't call here

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 28]
```diff
@@ -98,6 +111,114 @@ public void waitGraphsStarted() {
...
+        g.clearBackend();
+        try {
+            g.close();
+        } catch (Exception e) {
+            LOG.warn("Failed to close graph", e);
+        }
+        // Inject into gremlin server context
+        this.eventHub.call(Events.GRAPH_DROP, name);
+        // Inject into rest server context
+        this.graphs.remove(name);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:216]: use event

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 29]
```diff
@@ -98,6 +111,114 @@ public void waitGraphsStarted() {
...
+    }
+
+    private void checkOptions(HugeConfig config) {
+        // The store cannot be the same as the existing graph
+        this.checkOptionsUnique(config, CoreOptions.STORE);
+        // NOTE: rocksdb can't use same data path for different graph,
+        // but it's not easy to check here
+        String backend = config.get(CoreOptions.BACKEND);
+        if (backend.equalsIgnoreCase("rocksdb")) {
+            // TODO: should check data path...
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:197]: seem same data path is ok

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 30]
```diff
@@ -98,6 +111,114 @@ public void waitGraphsStarted() {
...
+    }
+
+    private HugeGraph createGraph(HugeConfig config) {
+        // open succeed will fill graph instance into HugeFactory graphs(map)
+        HugeGraph graph = (HugeGraph) GraphFactory.open(config);
+        if (this.requireAuthentication()) {
+            if (graph instanceof HugeGraphAuthProxy) {
+                HugeGraphAuthProxy proxy = (HugeGraphAuthProxy) graph;
+                // TODO：verfiyPermission is private
+                // proxy.verifyPermission(HugePermission.WRITE, ResourceType.STATUS);
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:172]: can call graph.mode(NONE) instead, but need some comments

**Review by javeme — CHANGES_REQUESTED** [review]

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 31]
```diff
@@ -98,6 +111,114 @@ public void waitGraphsStarted() {
...
+        return this.createGraph(cloneConfig);
+    }
+
+    public HugeGraph createGraph(String name, String configText) {
+        E.checkArgumentNotNull(name, "The graph name can't be null");
+        E.checkArgument(!this.graphs().contains(name),
+                        "The graph name '%s' has existed", name);
+
+        PropertiesConfiguration propConfig = this.buildConfig(configText);
+        HugeConfig config = new HugeConfig(propConfig, false);
```
Comment (dyhyao6) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:147]: Cannot resolve constructor 'HugeConfig(org.apache.commons.configuration.PropertiesConfiguration, boolean);
are you sure? @ @javeme @Linary

Comment (Linary) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:147]: This constructor was defined in hugegraph-common 1.7.10, it's not yet published to the maven central repository.

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java (hunk 29 above)
Comment (Linary) [reply hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:221]: same data path is not ok

**Comment by dyhyao6** [comment]
能不能加快实现图的热更新，thanks @Linary @javeme

File: hugegraph-dist/src/assembly/static/bin/start-hugegraph.sh [hunk 32]
```diff
@@ -57,8 +57,8 @@ check_port "$REST_SERVER_URL"
 
 echo "Starting HugeGraphServer..."
 
-"$BIN"/hugegraph-server.sh "$CONF"/gremlin-server.yaml "$CONF"/rest-server.properties \
-"$OPEN_SECURITY_CHECK" "$USER_OPTION" "$GC_OPTION" >>"$LOGS/hugegraph-server.log" 2>&1 &
+${BIN}/hugegraph-server.sh ${CONF}/gremlin-server.yaml ${CONF}/rest-server.properties \
```
Comment (javeme) [inline hugegraph-dist/src/assembly/static/bin/start-hugegraph.sh:60]: mkdir -p ${LOGS}

**Review by javeme — DISMISSED** [review]

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java [hunk 33]
```diff
@@ -100,6 +105,51 @@ public Object get(@Context GraphManager manager,
         return ImmutableMap.of("name", g.name(), "backend", g.backend());
     }
 
+    @PUT
+    @Timed
+    @Path("{name}")
+    @Consumes(MediaType.TEXT_PLAIN)
+    @Produces(APPLICATION_JSON_WITH_CHARSET)
+    @RolesAllowed({"admin", "$owner=$name"})
+    public Object manage(@Context GraphManager manager,
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java:114]: add delete method and move clone to create

File: hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java [hunk 34]
```diff
@@ -143,9 +130,18 @@ public Object manage(@Context GraphManager manager,
     @RolesAllowed({"admin"})
     public Object create(@Context GraphManager manager,
                          @PathParam("name") String name,
+                         @QueryParam("clone_graph_name") String clone,
                          String configText) {
-        LOG.debug("Create graph {} with config options '{}'", name, configText);
-        HugeGraph graph = manager.createGraph(name, configText);
+        HugeGraph graph;
+        if (clone != null && clone.isEmpty()) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/api/profile/GraphsAPI.java:136]: should be "!clone.isEmpty()"?

**Review by javeme — APPROVED** [review]

File: hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java [hunk 35]
```diff
@@ -108,6 +153,111 @@ public void waitGraphsStarted() {
...
+            throw e;
+        }
+        // Let gremlin server and rest server context add graph
+        this.eventHub.notify(Events.GRAPH_CREATE, graph);
+        // Write config to disk file
+        ConfigUtil.writeToFile(this.graphsDir, graph.name(), config);
+        return graph;
+    }
+
+    private PropertiesConfiguration buildConfig(String configText) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/com/baidu/hugegraph/core/GraphManager.java:219]: check configText not null? and seems empty string is ok

**Review by javeme — APPROVED** [review]

**Review by imbajin — APPROVED** [review]
test in client side
